795 delete repository asynchonously

eugenk opened the pull request:
Shall fix #795 and #342.

I created this branch on top of #1026, not on top of staging.

ebolloff commented on app/models/repository/destroying.rb:
```diff
@@ -2,28 +2,34 @@ module Repository::Destroying
-  def unmark_as_destroying
-    self.class.instance_variable_get(:@destroying).delete(self.id) if self.id
+  def can_be_deleted?
+    ontologies.map(&:can_be_deleted_with_whole_repository?).all?
```
Why don't you use `each` so you don't have to run over all ontologies?

eugenk replied:
Good point. I seem to have been in lazy Haskell mode.

ebolloff commented on app/views/repository_settings/_subnav.html.haml:
```diff
@@ -3,8 +3,14 @@
+      - else
+        .btn.btn-danger.tooltip-btn{ disabled: true, data: { toggle: 'tooltip', :'original-title' => t('repository.delete_error'), container: 'body' } }
+          %span.allow-tooltips
+            =t('repository.delete')
```
Please insert whitespace after `=`

ebolloff commented on app/controllers/ontology_search_controller.rb:
```diff
@@ -39,4 +39,8 @@ def refine_search_response
   end
+
+  def repository
+    Repository.find_by_path params[:repository_id]
```
Please use `()` around the parameter

ebolloff commented on app/models/repository/destroying.rb:
```diff
@@ -2,28 +2,34 @@ module Repository::Destroying
+  # It prepares the deletion by setting a flag, which enables the deletion
+  # of its ontologies.
   def destroy
+    Rails.logger.info "Destroy #{self.class} #{self} (id: #{id})"
```
Please use `()` around parameter

ebolloff commented on app/models/repository/destroying.rb:
```diff
@@ -2,28 +2,34 @@ module Repository::Destroying
+    unless can_be_deleted?
+      raise Repository::DeleteError, I18n.t('repository.delete_error')
+    end
+    Rails.logger.info "Mark #{self.class} #{self} (id: #{id}) as is_destroying"
```
Please use `()` around parameter

ebolloff commented on app/models/repository/destroying.rb:
```diff
@@ -2,28 +2,34 @@ module Repository::Destroying
-    raise e.class, "Can't delete repository: It contains an #{Settings.OMS} that is imported by another repository."
+    self.is_destroying = false
+    save!
+    raise e.class, I18n.t('repository.delete_error', oms: Settings.OMS)
```
Please instantiate a new error instance and set `e` as the origin so it won't get lost if you raise a new error

ebolloff commented:
Please rebase staging

ebolloff commented:
This won't fix #342 completely, because the repository seems to be deleted directly without marked as deleted. Please change the discription of the pull request, so it only fixes #795.

ebolloff commented:
This pull request should fix #342. Should the repository be marked as deleted in the gui or just in the model? Should there be a delete button for the owner, admin, ...?

ebolloff commented:
The repository should be marked as deleted in the gui for the admins & owner and there should be a `restore` button for them.

eugenk commented:
I rebased staging and added the restore funtionality. The first new commit is "Allow owners to see their deleted repositories." (6b4a7212feaabc9f8498c2e95f72d87b908c1154)

ebolloff commented:
:+1: after you used `save!`

eugenk commented:
If no one minds, I ignore the latest hound commits. Something seems to be wrong with the checker.

eugenk commented:
Travis fails because of different behaviour in the current version of hets. See https://github.com/spechub/Hets/issues/1455
